[LANGPLAT-87] Implement automated release tagging for dd-trace-go v2 (#3159)
DataDog/dd-trace-go: tools/autoreleasetagger/README.md

> # Auto Release Tagger

Reviewer: Up top can we also specify when you would use this tool? It's during the release process? And what is this new tool affording us that we did not have previously (with the old release process)?

> This tool ensures that all modules in a monorepo are tagged correctly, respecting dependency order—tagging dependencies first before their dependents.

Reviewer: I found this sentence hard to read, maybe you mean this instead?
```suggestion
This tool ensures that all modules in a monorepo are tagged correctly, respecting dependency order by tagging dependencies first before their dependents.
```
?

> This tool ensures that all modules in a monorepo are tagged correctly, respecting dependency order by tagging dependencies first before their dependents.

Reviewer 2: ```suggestion
This tool ensures that all modules in dd-trace-go are tagged correctly for release, respecting dependency order by tagging dependencies first before their dependents.

The repository has a structure consisting of nested modules for all the contribs and some packages to reduce the dependency surface to the strictly required dependencies.
```

- **Automatic Versioning**: Determines the next version based on existing tags.
- **Nested Module Support**: Tags all modules in a repository.
- **Dependency Awareness**: Ensures dependencies are tagged before their dependents.

## 📌 Prerequisites

Before running the tool, ensure you have updated the version according to the [Release Checklist](https://datadoghq.atlassian.net/wiki/spaces/DL/pages/2477949158/Template+v+MAJOR+.+MINOR+.+PATCH+Release+Checklist#Release-branch).

## 🛠️ Usage

Run the following command to tag the release:

```sh
go run ./tools/autoreleasetagger -root ../..
```

> Dry run:

Reviewer: Would be nice to see a quick statement about when/why you would do a Dry Run instead of a regular run, so that a reader can decide the correct use case for this command.
e.g.
`Dry run: Simulate the tagging process to check for potential issues`

> Dry run:

Reviewer 2: ```suggestion
Dry run: Simulate the tagging process to check for potential issues
```

```sh
go run ./tools/autoreleasetagger -dry-run -root ../..
```

> Run without pushing tags:

Reviewer: Same comment as Dry Run. Admittedly, commenting that because I at first had to think critically about why you might want to do those operations.

> Run without pushing tags:

Reviewer 2: ```suggestion
Run without pushing tags: Tag all the modules to release but don't push them
```
